### ISSUE TITLE
PostgreSQL wire protocol: type OID for text responses in prepared statements is invalid

### ISSUE THREAD

**Issue by emarsden** [crate/crate#16500]
### CrateDB version

5.8.1

### CrateDB setup information

1 node, basic configuration using the official Docker/Podman container.

### Problem description

A text field in a response to a prepared statement is returned with the type oid of 705 ("unknown" in pg_type), rather than the expected value 25 ("text"), or the value 1043 ("varchar") which is used in a response for the same query using the simple query protocol.

Using a valid type OID is important because clients may decide to parse (and for example do apply Unicode decoding) depending on the type OID. 

### Steps to Reproduce

1. Make a Prepare/Bind/Execute query of the form "SELECT $1" with argument "foo" of type TEXT.
2. The RowDescription message in the response has a type OID of 705, rather than expected 25 ("text"). 

### Actual Result

GitHub won't allow me to attach a pcap file, but attached are two Wireshark screenshots from a capture, the first showing the invalid "Type OID: 705" in a response to a P/B/E query, and the second showing the valid "Type OID: 1043" (which corresponds to "varchar" and is different from PostgreSQL which uses 25="text").

![pcap-valid](https://github.com/user-attachments/assets/836d05c5-3311-4ae3-b596-cd0a7b5adff3)
![pcap-invalid](https://github.com/user-attachments/assets/91564816-fd80-49a9-b03e-6341e881e57d)

### Expected Result

Expected Type OID of 25 or Type OID of 1043 as for a query using the standard query protocol.

**Comment by mfussenegger** [comment]
Does the parse message include the type information?
The problem is that in `select $1` there's otherwise no way to infer the type, that's why CrateDB default's to unknown.

I guess what we could do is run another type guess after we've received the value from the bind message, but not sure if that's a real improvement, given that the value was decoded using the unknown type.


Do you have a bit more context on how this came up? Is there a client where this causes problems?

**Comment by emarsden** [comment]
Yes, the Parse message in the extended query protocol specifies the OID (it is set to 25 = "text" in this case), so it shouldn't be decoded as the unknown type.

Using select $1::text returns an OID of 1043 ("varchar"), which is different behaviour from PostgreSQL but seems OK.

Context: I am the author of a [database client library](https://github.com/emarsden/pg-el) that implements the PostgreSQL wire protocol and works partially with CrateDB. It uses the per-column returned OID to determine how values should be deserialized, and this CrateDB behaviour means that text fields are not deserialized correctly.

**Comment by mfussenegger** [comment]
Could you run the following two commands to change the log levels, and then re-run your scenario and post the log output?

```
set global "logger.io.crate.action.sql" = 'TRACE';
set global "logger.io.crate.protocols.postgres" = 'TRACE';
```

As far as I can tell with a unit test, we also return `1043` if the parse message includes a `25` or `1043` oid as parameter type hint.
(that it returns 1043 instead of 25 may seem a bit odd, but that's because we internally use a single string type, so we loose the detail that it was input as 25)

**Comment by emarsden** [comment]
OK, I have run your test, output below. What I wrote earlier concerning the type in the parse message was incorrect (my apologies); my library was in fact specifying Type OID = 0. With a Type OID = 25 the behaviour of CrateDB is OK (modulo 25 != 1043).

However, with the same query and Type OID = 0 in the Parse message, PostgreSQL infers that the output type is text and returns Type OID = 25 (see attached Wireshark screenshots). The difference for a client is that it will know it should apply the client encoding for OID=25 or OID=1043, but not if OID=unknown, leading to Unicode errors.

![postgresql-S1](https://github.com/user-attachments/assets/2854d378-b5b8-4da5-b8b5-6a0233b8a3d7)
![postgresql-S2](https://github.com/user-attachments/assets/0f1285eb-327a-4905-b5b4-9834494adfd1)

```
[TRACE][i.c.p.p.PostgresWireProtocol] [Jungfrau] msg=P msgLength=17 readableBytes=17
[DEBUG][i.c.a.s.Sessions         ] [Jungfrau] method=parse stmtName= query=SELECT $1 paramTypes=[undefined]
[TRACE][i.c.p.p.PostgresWireProtocol] [Jungfrau] msg=B msgLength=21 readableBytes=21
[DEBUG][i.c.a.s.Sessions         ] [Jungfrau] method=bind portalName= statementName= params=[fooblé]
[TRACE][i.c.p.p.PostgresWireProtocol] [Jungfrau] msg=D msgLength=2 readableBytes=2
[DEBUG][i.c.a.s.Sessions         ] [Jungfrau] method=describe type=P portalOrStatement=
[TRACE][i.c.p.p.PostgresWireProtocol] [Jungfrau] msg=E msgLength=5 readableBytes=5
[DEBUG][i.c.a.s.Sessions         ] [Jungfrau] method=execute portalName= maxRows=0
[TRACE][i.c.p.p.PostgresWireProtocol] [Jungfrau] msg=S msgLength=0 readableBytes=0
[TRACE][i.c.p.p.Messages         ] [Jungfrau] sentParseComplete
[TRACE][i.c.p.p.Messages         ] [Jungfrau] sentBindComplete
[TRACE][i.c.p.p.Messages         ] [Jungfrau] sentRowDescription
[TRACE][i.c.p.p.Messages         ] [Jungfrau] sentCommandComplete
[TRACE][i.c.p.p.Messages         ] [Jungfrau] sentReadyForQuery
```
